crates/autopilot/src/solvable_orders.rs

```diff
@@ -156,8 +158,95 @@ impl SolvableOrdersCache {
         );
     }
 
-    pub fn current_auction(&self) -> Option<domain::Auction> {
-        self.cache.lock().unwrap().auction.clone()
+    pub async fn current_auction(&self) -> Option<domain::Auction> {
+        self.cache
+            .lock()
+            .await
+            .as_ref()
+            .map(|inner| inner.auction.clone())
+    }
+
+    fn build_solvable_orders(
+        current_orders: &boundary::SolvableOrders,
+        new_orders: Vec<database::orders::OrderWithoutTrades>,
+        mut new_trades: HashMap<domain::OrderUid, database::trades::TradedAmounts>,
+        new_quotes: HashMap<domain::OrderUid, domain::Quote>,
+    ) -> Result<boundary::SolvableOrders> {
+        let now = now_in_epoch_seconds();
+        let mut orders = current_orders.orders.clone();
+        // Remove expired orders.
+        orders.retain(|_uid, order| order.data.valid_to >= now);
+        let mut quotes = current_orders.quotes.clone();
+        let mut latest_trade_block = current_orders.latest_settlement_block;
+
+        for new_order in new_orders {
```

Comment:
This logic seems pretty hard to grasp.
Could we first try to implement a simpler approach in 2 phases?
1. blindly add new orders and update existing orders based on trades
2. go over all orders and filter out all the expired, filled, etc. ones

Since caching is known for being tricky I would only consider more optimal implementations if we definitely need the extra performance.

Reply:
IMO, that looks even more complicated. Since we don't have trades data for new orders, the process will look as follows:
1. Create new full orders based on the trades data and insert them into the cache.
2. Iterate over the fetched orders again to find and remove expired/fulfilled/canceled orders from the cache.
3. Iterate over the cache to do the same with the old existing orders.

Reply:
IMO, that looks even more complicated. Since we don't have trades data for new orders, the process will look as follows:

1. Create new full orders based on the trades data and insert them into the cache.
2. Iterate over the fetched orders(that don't have associated trades data) again to find and remove expired/fulfilled/canceled orders from the cache.
3. Iterate over the cache to do the same with the old existing orders.

Reply:
> Also why do we need to have trades separately? 

I was thinking about this today and something stopped me from doing this. But after reverting to that one more time, I don't really see any blocker.

Reply:
Damn, failing the test reminded me what the issue was initially. Recent trades might contain data for orders created before the latest timestamp. Will revert the latest changes...

Reply:
It turned out that not only trade data need to be fetched separately. Update the PR description.

Reply:
> Having said that, the data needs to be fetched within a single TX, so there is no way to run all the queries in parallel.

> Recent trades might contain data for orders created before the latest timestamp.

Those 2 issues could still be resolved by the simpler approach, no?
I was imagining a single (probably somewhat complex) query that just returns the list of uids where **something** changed since last time. This can be parallelized as postgres sees fit.
Afterwards use the eixisting query to retrieve "full orders" from the DB for each uid, add/replace them in the cache and finally filter out all the orders that are not desirable.

Would this not make things easier?

Reply:
Found another potential corner case.
Since [onchain events are split into chunks](https://github.com/cowprotocol/services/blob/e7de6bfd0526e00b55fda817b33a0fde715c8a61/crates/shared/src/event_handling.rs#L283) with hardcoded value, there is a possibility for a case when a data for one block is separated into different chunks and order updates fetched in between these two operations making the logic store block X as the last seen, but in fact, there are remaining records.